buildkitd/certificates.go

```diff
@@ -11,53 +11,160 @@ import (
+
+	typeCert   = "CERTIFICATE"
+	typeRSAKey = "RSA PRIVATE KEY"
 )
 
-// GenerateCertificates creates and saves a CA and certificates for both sides of an mTLS TCP connection.
-func GenerateCertificates(dir, hostname string) error {
-	ca, err := createAndSaveCA(dir)
-	if err != nil {
-		return errors.Wrap(err, "create CA")
+// GenCerts creates and saves a CA and certificates for both sides of an mTLS TCP connection.
+func GenCerts(cfg config.Config, hostname string) error {
+	caKey, err := parseTLSKey(cfg.Global.TLSCAKey)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return errors.Wrap(err, "failed reading CA key")
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		all := []string{
+			cfg.Global.TLSCACert,
+			cfg.Global.ServerTLSCert,
+			cfg.Global.ServerTLSKey,
+			cfg.Global.ClientTLSCert,
+			cfg.Global.ClientTLSKey,
+		}
+		var missing []string
+		for _, f := range all {
+			if exists, _ := fileutil.FileExists(f); !exists {
+				missing = append(missing, f)
+			}
+		}
+		switch len(missing) {
+		case 0:
+			return nil
+		case len(all):
+			key, err := createTLSKey(cfg.Global.TLSCAKey)
+			if err != nil {
+				return errors.Wrap(err, "could not create CA")
+			}
+			caKey = key
+		default:
+			found := all
+			for _, m := range missing {
+				for i, f := range found {
+					if f == m {
+						found = append(found[:i], found[i+1:]...)
+						break
+					}
+				}
+			}
+			return hint.Wrap(errors.New("cannot generate missing certificates"),
+				fmt.Sprintf("missing certificates: %v", missing),
+				fmt.Sprintf("found certificates: %v", found),
+				"you may want to stop earthly-buildkitd, delete your certificates, and run 'earthly bootstrap' to regenerate certificates",
+			)
+		}
+	}
+	caCert, err := parseTLSCert(cfg.Global.TLSCACert)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return errors.Wrap(err, "could not parse CA certificate")
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		caCert, err = createCACert(caKey, cfg.Global.TLSCACert)
+		if err != nil {
+			return errors.Wrap(err, "could not create CA certificate")
+		}
+	}
+	ca := &certData{
+		Key:  caKey,
+		Cert: caCert,
+	}
+
+	if err := genCert(ca, buildkit, hostname, cfg.Global.ServerTLSKey, cfg.Global.ServerTLSCert); err != nil {
+		return errors.Wrapf(err, "could not generate server TLS key/cert pair for %v", buildkit)
+	}
+	if err := genCert(ca, earthly, hostname, cfg.Global.ClientTLSKey, cfg.Global.ClientTLSCert); err != nil {
+		return errors.Wrapf(err, "could not generate client TLS key/cert pair for %v", earthly)
 	}
 
-	err = createAndSaveCertificate(ca, buildkit, dir, hostname)
-	if err != nil {
-		return errors.Wrap(err, "create buildkit certificate")
+	return nil
+}
+
+func genCert(ca *certData, role, hostname, keyPath, certPath string) error {
+	key, err := parseTLSKey(keyPath)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return errors.Wrapf(err, "could not parse %v TLS key", role)
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		os.Remove(certPath)
```

Comment:
this feels a big dangerous, since a user might have typoed an existing key path, then their cert gets deleted.

what if instead we only generated a new key+cert when both keyPath and certPath don't exist?